### PR TITLE
texture_cache: Split out template definitions

### PR DESCRIPTION
**Pull request by yzct12345**
It takes as much as three minutes to incrementally compile yuzu when I edit any of the template functions in `texture_cache.h`. It is compiling about 20 other unrelated files. This was already frustrating for me when writing #6790 and it gets worse debugging trickier problems. By splitting out the definitions the compile merely takes 30 seconds.

### CODE REVIEW

File: src/video_core/CMakeLists.txt [hunk 1]
```diff
@@ -186,6 +188,7 @@ add_library(video_core STATIC
     texture_cache/samples_helper.h
     texture_cache/slot_vector.h
     texture_cache/texture_cache.h
+    texture_cache/texture_cache_templates.h
```
Comment (ameerj) [inline src/video_core/CMakeLists.txt:191]: Do you mind keeping the old `texture_cache.h` and adding a `texture_cache_base.h` which holds the few hundred line of code that were not moved to the templates.h? Makes the diff more manageable.

**Comment by goldenx86** [comment]
You need moar coars.

File: src/video_core/CMakeLists.txt (hunk 1 above)
Comment (yzct12345) [reply src/video_core/CMakeLists.txt]: Done. In the new diff it's no longer obvious I removed the "InvalidateColorBuffer" "InvalidateDepthBuffer" methods that don't compile.

File: src/video_core/texture_cache/texture_cache_base.h [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+#include "video_core/dirty_flags.h"
+#include "video_core/engines/fermi_2d.h"
+#include "video_core/engines/kepler_compute.h"
+#include "video_core/engines/maxwell_3d.h"
+#include "video_core/memory_manager.h"
+#include "video_core/rasterizer_interface.h"
+#include "video_core/surface.h"
+#include "video_core/texture_cache/descriptor_table.h"
+#include "video_core/texture_cache/format_lookup_table.h"
+#include "video_core/texture_cache/formatter.h"
```
Comment (ameerj) [inline src/video_core/texture_cache/texture_cache_base.h:38]: I'm pretty sure most of these headers are not needed for `texture_cache_base.h`. I would appreciate if you can minimize the number of headers and `using` directives that are declared in this file, and move them back to `texture_cache.h` if they're still being used.

File: src/video_core/texture_cache/texture_cache_base.h [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+#pragma once
+
+#include <array>
+#include <mutex>
+#include <span>
+#include <type_traits>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
```
Comment (ameerj) [inline src/video_core/texture_cache/texture_cache_base.h:13]: I don't see this being used in this header file

File: src/video_core/texture_cache/texture_cache_base.h [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+#include <vector>
+
+#include <boost/container/small_vector.hpp>
```
Comment (ameerj) [inline src/video_core/texture_cache/texture_cache_base.h:16]: I don't see this being used in this header file

File: src/video_core/texture_cache/texture_cache_base.h [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+#include "common/alignment.h"
```
Comment (ameerj) [inline src/video_core/texture_cache/texture_cache_base.h:18]: I don't see this being used in this header file

File: src/video_core/texture_cache/texture_cache_base.h [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+#include "common/common_types.h"
+#include "common/literals.h"
+#include "common/settings.h"
```
Comment (ameerj) [inline src/video_core/texture_cache/texture_cache_base.h:21]: I don't see this being used in this header file

File: src/video_core/texture_cache/texture_cache_base.h [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+#include "video_core/texture_cache/render_targets.h"
+#include "video_core/texture_cache/slot_vector.h"
+#include "video_core/texture_cache/types.h"
+#include "video_core/texture_cache/util.h"
+#include "video_core/textures/texture.h"
+
+namespace VideoCommon {
+
+using Tegra::Texture::SwizzleSource;
+using Tegra::Texture::TextureType;
```
Comment (ameerj) [inline src/video_core/texture_cache/texture_cache_base.h:43]: I don't see this being used in this header file

File: src/video_core/texture_cache/texture_cache_base.h [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+using Tegra::Texture::TICEntry;
+using Tegra::Texture::TSCEntry;
+using VideoCore::Surface::GetFormatType;
+using VideoCore::Surface::IsCopyCompatible;
+using VideoCore::Surface::PixelFormat;
+using VideoCore::Surface::PixelFormatFromDepthFormat;
+using VideoCore::Surface::PixelFormatFromRenderTargetFormat;
+using VideoCore::Surface::SurfaceType;
```
Comment (ameerj) [inline src/video_core/texture_cache/texture_cache_base.h:51]: I don't see this being used in this header file.

**Review by ameerj — COMMENTED** [review]
Just at a quick glance I saw these aren't being used. I'd suggest reviewing and minimizing all the includes here to make this header as light as possible (considering the spirit of this PR is to minimize compilation time)

**Review by ameerj — COMMENTED** [review]
LGTM. Gonna EA tag this for a bit to make sure no compiler issues pop up.

**Pull request merged**